IMap.destroy() - map is destroyed and then immediately recreated

Creating this issue as suggested by @bilalyasar (see [discussion on the Hazelcast Google Groups forum] (https://groups.google.com/forum/#!topic/hazelcast/G1P43tORJPA))

Hazelcast version that you use: ```3.4.2```, ```3.5```
Cluster size: ```1```
Number of the clients: ```1```
Version of Java: ```OracleJDK 1.7.0_80, x86_64```, ```OracleJDK 1.8.0_45, x86_64```
Operating system: ```OS X Yosemite 10.10.3 (14D136)```
Logs and stack traces: See the console output from executing the TestNG test case below
Detailed description of the steps to reproduce your issue: See the TestNG test case below
Unit test with the hazelcast.xml file: 

```java
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotNull;

import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

import com.hazelcast.client.HazelcastClient;
import com.hazelcast.core.DistributedObjectEvent;
import com.hazelcast.core.DistributedObjectListener;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;

public class IMapDestroyTest {

	private HazelcastInstance server;
	private HazelcastInstance client;

	@BeforeClass
	public void setUpBeforeClass() {
		server = Hazelcast.newHazelcastInstance();
		sleep(2000);
		client = HazelcastClient.newHazelcastClient();
		client.addDistributedObjectListener(new DistributedObjectListener() {
			@Override public void distributedObjectDestroyed(final DistributedObjectEvent event) {
				System.out.println("\n\tdistributedObjectDestroyed(): " + event);
			}
			@Override public void distributedObjectCreated(final DistributedObjectEvent event) {
				System.out.println("\n\tdistributedObjectCreated(): " + event);
			}
		});
	}

	@AfterClass
	public void tearDownAfterClass() {
		client.shutdown();
		server.shutdown();
	}

	@Test
	public void testIMapDestroy() {
		System.out.print("Creating 'foo' map... ");
		IMap<String, String> fooMap = client.getMap("foo");
		assertNotNull(fooMap);
		System.out.println("done.");
		System.out.println("All known distributed objects: " + client.getDistributedObjects());
		System.out.print("Inserting an entry... ");
		fooMap.put("fu", "bar");
		System.out.print("done.\nRetrieving the entry... ");
		String value = fooMap.get("fu");
		assertEquals(value, "bar");
		System.out.print("done.\nDestroying the map... ");
		fooMap.destroy();
		System.out.println("done.");
		System.out.println("All known distributed objects: " + client.getDistributedObjects());
	}

	private void sleep(final int millis) {
		try {
			Thread.sleep(millis);
		} catch (InterruptedException e) {
			Thread.currentThread().interrupt();
		}
	}
}
```

Console output:
```
Creating 'foo' map...
        distributedObjectCreated(): DistributedObjectEvent{eventType=CREATED, serviceName='hz:impl:mapService', distributedObject=IMap{name='foo'}}
done.
All known distributed objects: [IMap{name='foo'}]
Inserting an entry... done.
Retrieving the entry... done.
Destroying the map...
        distributedObjectDestroyed(): DistributedObjectEvent{eventType=DESTROYED, serviceName='hz:impl:mapService', distributedObject=IMap{name='foo'}}
        distributedObjectCreated(): DistributedObjectEvent{eventType=CREATED, serviceName='hz:impl:mapService', distributedObject=IMap{name='foo'}}
done.
All known distributed objects: [IMap{name='foo'}]
```

I've just tested with the lastest Hazelcast version ```3.5.1``` and this test still fails. My test is now slightly different from the one I originally posted (it now uses lifecycle event listener to detect server and client start-up, instead of sleeping), but I tested with both and both fail. Here's the code of my latest test:

```java
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotNull;
import static org.testng.Assert.assertTrue;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

import com.hazelcast.client.HazelcastClient;
import com.hazelcast.client.config.ClientConfig;
import com.hazelcast.client.config.XmlClientConfigBuilder;
import com.hazelcast.config.Config;
import com.hazelcast.config.ListenerConfig;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.DistributedObjectEvent;
import com.hazelcast.core.DistributedObjectListener;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import com.hazelcast.core.LifecycleEvent;
import com.hazelcast.core.LifecycleEvent.LifecycleState;
import com.hazelcast.core.LifecycleListener;

public class IMapDestroyTest {

	private HazelcastInstance server;
	private HazelcastInstance client;

	@BeforeClass
	public void setUpBeforeClass() throws InterruptedException {
		final CountDownLatch signalStarted = new CountDownLatch(2);

		server = startServer(signalStarted);
		client = startClient(signalStarted);

		signalStarted.await(15, TimeUnit.SECONDS);
	}

	private HazelcastInstance startServer(final CountDownLatch signalStarted) {
		return Hazelcast.newHazelcastInstance(serverConfig(signalStarted));
	}

	private Config serverConfig(final CountDownLatch signalStarted) {
		Config config = new XmlConfigBuilder().build();
		config.addListenerConfig(new ListenerConfig(new StartedListener(signalStarted)));
		return config;
	}

	private HazelcastInstance startClient(final CountDownLatch signalStarted) {
		HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig(signalStarted));

		client.addDistributedObjectListener(new DistributedObjectListener() {
			@Override public void distributedObjectDestroyed(final DistributedObjectEvent event) {
				System.out.println("distributedObjectDestroyed(): " + event);
			}
			@Override public void distributedObjectCreated(final DistributedObjectEvent event) {
				System.out.println("distributedObjectCreated(): " + event);
			}
		});

		return client;
	}

	private ClientConfig clientConfig(final CountDownLatch signalStarted) {
		ClientConfig clientConfig = new XmlClientConfigBuilder().build();
		clientConfig.addListenerConfig(new ListenerConfig(new StartedListener(signalStarted)));
		return clientConfig;
	}

	@AfterClass
	public void tearDownAfterClass() {
		client.shutdown();
		server.shutdown();
	}

	@Test
	public void testIMapDestroy() {
		assertTrue(client.getDistributedObjects().isEmpty(), "There should be no distributed objects on a fresh start");

		IMap<String, String> fooMap = client.getMap("foo");

		assertNotNull(fooMap);
		assertEquals(client.getDistributedObjects().size(), 1, "There should be one distributed object - the 'foo' IMap");

		fooMap.put("fu", "bar");
		String value = fooMap.get("fu");

		assertEquals(value, "bar");

		fooMap.destroy();

		assertTrue(client.getDistributedObjects().isEmpty(), "There should be no distributed objects after destroying the map");
	}

	private static class StartedListener implements LifecycleListener {
		private final CountDownLatch signalStarted;

		public StartedListener(final CountDownLatch signalStarted) {
			this.signalStarted = signalStarted;
		}

		@Override
		public void stateChanged(final LifecycleEvent event) {
			if (event.getState() == LifecycleState.STARTED) {
				signalStarted.countDown();
			}
		}
	}
}
```

And this is the test output:

```
distributedObjectCreated(): DistributedObjectEvent{eventType=CREATED, serviceName='hz:impl:mapService', distributedObject=IMap{name='foo'}}
distributedObjectDestroyed(): DistributedObjectEvent{eventType=DESTROYED, serviceName='hz:impl:mapService', distributedObject=IMap{name='foo'}}
distributedObjectCreated(): DistributedObjectEvent{eventType=CREATED, serviceName='hz:impl:mapService', distributedObject=IMap{name='foo'}}
FAILED: testIMapDestroy
java.lang.AssertionError: There should be no distributed objects after destroying the map expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:94)
	at org.testng.Assert.failNotEquals(Assert.java:496)
	at org.testng.Assert.assertTrue(Assert.java:42)
	at my.experiments.hazelcast.IMapDestroyTest.testIMapDestroy(IMapDestroyTest.java:91)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:659)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:845)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1153)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:771)
	at org.testng.TestRunner.run(TestRunner.java:621)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1199)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1124)
	at org.testng.TestNG.run(TestNG.java:1032)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)
```

Does this work for you?

--- a/hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/ProxyManager.java
+++ b/hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/ProxyManager.java
@@ -25,8 +25,8 @@ import com.hazelcast.client.config.ClientProperties;
 import com.hazelcast.client.config.ProxyFactoryConfig;
 import com.hazelcast.client.connection.ClientConnectionManager;
 import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
-import com.hazelcast.client.impl.client.ClientCreateRequest;
 import com.hazelcast.client.impl.client.AddDistributedObjectListenerRequest;
+import com.hazelcast.client.impl.client.ClientCreateRequest;
 import com.hazelcast.client.impl.client.RemoveDistributedObjectListenerRequest;
 import com.hazelcast.client.proxy.ClientAtomicLongProxy;
 import com.hazelcast.client.proxy.ClientAtomicReferenceProxy;
@@ -45,6 +45,7 @@ import com.hazelcast.client.proxy.ClientSemaphoreProxy;
 import com.hazelcast.client.proxy.ClientSetProxy;
 import com.hazelcast.client.proxy.ClientTopicProxy;
 import com.hazelcast.client.spi.impl.ClientInvocation;
+import com.hazelcast.client.spi.impl.listener.LazyDistributedObjectEvent;
 import com.hazelcast.client.txn.proxy.xa.XAResourceProxy;
 import com.hazelcast.collection.impl.list.ListService;
 import com.hazelcast.collection.impl.queue.QueueService;
@@ -81,7 +82,6 @@ import com.hazelcast.topic.impl.reliable.ReliableTopicService;
 import com.hazelcast.transaction.impl.xa.XAService;
 import com.hazelcast.util.EmptyStatement;
 import com.hazelcast.util.ExceptionUtil;
-
 import java.io.IOException;
 import java.lang.reflect.Constructor;
 import java.util.Collection;
@@ -337,35 +337,50 @@ public final class ProxyManager {
     }
 
     public String addDistributedObjectListener(final DistributedObjectListener listener) {
-        final AddDistributedObjectListenerRequest request = new AddDistributedObjectListenerRequest();
-        final EventHandler<PortableDistributedObjectEvent> eventHandler = new EventHandler<PortableDistributedObjectEvent>() {
-            public void handle(PortableDistributedObjectEvent e) {
-                final ObjectNamespace ns = new DefaultObjectNamespace(e.getServiceName(), e.getName());
-                ClientProxyFuture future = proxies.get(ns);
-                ClientProxy proxy = future == null ? null : future.get();
-                if (proxy == null) {
-                    proxy = getOrCreateProxy(e.getServiceName(), e.getName());
-                }
-
-                DistributedObjectEvent event = new DistributedObjectEvent(e.getEventType(), e.getServiceName(), proxy);
-                if (DistributedObjectEvent.EventType.CREATED.equals(e.getEventType())) {
-                    listener.distributedObjectCreated(event);
-                } else if (DistributedObjectEvent.EventType.DESTROYED.equals(e.getEventType())) {
-                    listener.distributedObjectDestroyed(event);
-                }
-            }
-
-            @Override
-            public void beforeListenerRegister() {
-            }
-
-            @Override
-            public void onListenerRegister() {
-
-            }
-        };
+        AddDistributedObjectListenerRequest request = new AddDistributedObjectListenerRequest();
+        EventHandler<PortableDistributedObjectEvent> eventHandler = new DistributedObjectEventHandler(listener, this);
         return client.getListenerService().registerListener(request, eventHandler);
     }
+
+
+    private final class DistributedObjectEventHandler implements EventHandler<PortableDistributedObjectEvent> {
+
+        private final DistributedObjectListener listener;
+        private ProxyManager proxyManager;
+
+        private DistributedObjectEventHandler(DistributedObjectListener listener, ProxyManager proxyManager) {
+            this.listener = listener;
+            this.proxyManager = proxyManager;
+        }
+
+        @Override
+        public void handle(PortableDistributedObjectEvent e) {
+            String name = e.getName();
+            String serviceName = e.getServiceName();
+            DistributedObjectEvent.EventType eventType = e.getEventType();
+            ObjectNamespace ns = new DefaultObjectNamespace(serviceName, name);
+            ClientProxyFuture future = proxies.get(ns);
+            ClientProxy proxy = future == null ? null : future.get();
+            LazyDistributedObjectEvent event = new LazyDistributedObjectEvent(eventType, serviceName, name, proxy,
+                    proxyManager);
+            if (DistributedObjectEvent.EventType.CREATED.equals(eventType)) {
+                listener.distributedObjectCreated(event);
+            } else if (DistributedObjectEvent.EventType.DESTROYED.equals(eventType)) {
+                listener.distributedObjectDestroyed(event);
+            }
+        }
+
+        @Override
+        public void beforeListenerRegister() {
+
+        }
+
+        @Override
+        public void onListenerRegister() {
+
+        }
+    }
+
 
     public boolean removeDistributedObjectListener(String id) {
         final RemoveDistributedObjectListenerRequest request = new RemoveDistributedObjectListenerRequest(id);

--- a/hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/LazyDistributedObjectEvent.java
+++ b/hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/LazyDistributedObjectEvent.java
@@ -1,0 +1,52 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.spi.impl.listener;
+
+import com.hazelcast.client.spi.ProxyManager;
+import com.hazelcast.core.DistributedObject;
+import com.hazelcast.core.DistributedObjectEvent;
+
+public final class LazyDistributedObjectEvent extends DistributedObjectEvent {
+
+    private ProxyManager proxyManager;
+
+    /**
+     * Constructs a DistributedObject Event.
+     *
+     * @param eventType         The event type as an enum {@link EventType} integer.
+     * @param serviceName       The service name of the DistributedObject.
+     * @param objectName        The name of the DistributedObject.
+     * @param distributedObject The DistributedObject for the event.
+     * @param proxyManager      The ProxyManager for lazily creating the proxy if not available on the client.
+     */
+    public LazyDistributedObjectEvent(EventType eventType, String serviceName, String objectName,
+                                      DistributedObject distributedObject,
+                                      ProxyManager proxyManager) {
+        super(eventType, serviceName, objectName, distributedObject);
+        this.proxyManager = proxyManager;
+    }
+
+    @Override
+    public DistributedObject getDistributedObject() {
+        distributedObject = super.getDistributedObject();
+        if (distributedObject == null) {
+            distributedObject = proxyManager.getOrCreateProxy(getServiceName(), (String) getObjectName());
+        }
+        return distributedObject;
+    }
+
+}

--- a/hazelcast-client-legacy/src/test/java/com/hazelcast/client/proxy/DistributedObjectListenerTest.java
+++ b/hazelcast-client-legacy/src/test/java/com/hazelcast/client/proxy/DistributedObjectListenerTest.java
@@ -17,20 +17,24 @@
 package com.hazelcast.client.proxy;
 
 import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.core.DistributedObject;
 import com.hazelcast.core.DistributedObjectEvent;
 import com.hazelcast.core.DistributedObjectListener;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.ITopic;
+import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import java.util.Collection;
+import java.util.concurrent.CountDownLatch;
 import org.junit.After;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.util.concurrent.CountDownLatch;
+import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
@@ -46,7 +50,7 @@ public class DistributedObjectListenerTest extends HazelcastTestSupport {
     @Test
     public void destroyedNotReceivedOnClient() throws Exception {
         HazelcastInstance instance = hazelcastFactory.newHazelcastInstance();
-        HazelcastInstance client = hazelcastFactory.newHazelcastClient();
+        final HazelcastInstance client = hazelcastFactory.newHazelcastClient();
         final CountDownLatch createdLatch = new CountDownLatch(1);
         final CountDownLatch destroyedLatch = new CountDownLatch(1);
         client.addDistributedObjectListener(new DistributedObjectListener() {
@@ -65,6 +69,13 @@ public class DistributedObjectListenerTest extends HazelcastTestSupport {
         assertOpenEventually(createdLatch, 10);
         topic.destroy();
         assertOpenEventually(destroyedLatch, 10);
+        assertTrueAllTheTime(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                Collection<DistributedObject> distributedObjects = client.getDistributedObjects();
+                assertTrue(distributedObjects.isEmpty());
+            }
+        }, 5);
     }
 
 }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/ProxyManager.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/ProxyManager.java
@@ -48,6 +48,7 @@ import com.hazelcast.client.proxy.ClientTopicProxy;
 import com.hazelcast.client.proxy.txn.xa.XAResourceProxy;
 import com.hazelcast.client.spi.impl.ClientInvocation;
 import com.hazelcast.client.spi.impl.ListenerMessageCodec;
+import com.hazelcast.client.spi.impl.listener.LazyDistributedObjectEvent;
 import com.hazelcast.collection.impl.list.ListService;
 import com.hazelcast.collection.impl.queue.QueueService;
 import com.hazelcast.collection.impl.set.SetService;
@@ -357,7 +358,7 @@ public final class ProxyManager {
     }
 
     public String addDistributedObjectListener(final DistributedObjectListener listener) {
-        final EventHandler<ClientMessage> eventHandler = new DistributedObjectEventHandler(listener);
+        final EventHandler<ClientMessage> eventHandler = new DistributedObjectEventHandler(listener, this);
         return client.getListenerService().registerListener(distributedObjectListenerCodec, eventHandler);
     }
 
@@ -365,9 +366,11 @@ public final class ProxyManager {
             implements EventHandler<ClientMessage> {
 
         private final DistributedObjectListener listener;
+        private ProxyManager proxyManager;
 
-        private DistributedObjectEventHandler(DistributedObjectListener listener) {
+        private DistributedObjectEventHandler(DistributedObjectListener listener, ProxyManager proxyManager) {
             this.listener = listener;
+            this.proxyManager = proxyManager;
         }
 
         @Override
@@ -375,13 +378,9 @@ public final class ProxyManager {
             final ObjectNamespace ns = new DefaultObjectNamespace(serviceName, name);
             ClientProxyFuture future = proxies.get(ns);
             ClientProxy proxy = future == null ? null : future.get();
-            if (proxy == null) {
-                proxy = getOrCreateProxy(serviceName, name);
-            }
-
             DistributedObjectEvent.EventType eventType = DistributedObjectEvent.EventType.valueOf(eventTypeName);
-            DistributedObjectEvent event = new DistributedObjectEvent(eventType,
-                    serviceName, proxy);
+            LazyDistributedObjectEvent event = new LazyDistributedObjectEvent(eventType, serviceName, name, proxy,
+                    proxyManager);
             if (DistributedObjectEvent.EventType.CREATED.equals(eventType)) {
                 listener.distributedObjectCreated(event);
             } else if (DistributedObjectEvent.EventType.DESTROYED.equals(eventType)) {

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/LazyDistributedObjectEvent.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/LazyDistributedObjectEvent.java
@@ -1,0 +1,52 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.spi.impl.listener;
+
+import com.hazelcast.client.spi.ProxyManager;
+import com.hazelcast.core.DistributedObject;
+import com.hazelcast.core.DistributedObjectEvent;
+
+public final class LazyDistributedObjectEvent extends DistributedObjectEvent {
+
+    private ProxyManager proxyManager;
+
+    /**
+     * Constructs a DistributedObject Event.
+     *
+     * @param eventType         The event type as an enum {@link EventType} integer.
+     * @param serviceName       The service name of the DistributedObject.
+     * @param objectName        The name of the DistributedObject.
+     * @param distributedObject The DistributedObject for the event.
+     * @param proxyManager      The ProxyManager for lazily creating the proxy if not available on the client.
+     */
+    public LazyDistributedObjectEvent(EventType eventType, String serviceName, String objectName,
+                                      DistributedObject distributedObject,
+                                      ProxyManager proxyManager) {
+        super(eventType, serviceName, objectName, distributedObject);
+        this.proxyManager = proxyManager;
+    }
+
+    @Override
+    public DistributedObject getDistributedObject() {
+        distributedObject = super.getDistributedObject();
+        if (distributedObject == null) {
+            distributedObject = proxyManager.getOrCreateProxy(getServiceName(), (String) getObjectName());
+        }
+        return distributedObject;
+    }
+
+}

--- a/hazelcast-client/src/test/java/com/hazelcast/client/proxy/DistributedObjectListenerTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/proxy/DistributedObjectListenerTest.java
@@ -17,20 +17,24 @@
 package com.hazelcast.client.proxy;
 
 import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.core.DistributedObject;
 import com.hazelcast.core.DistributedObjectEvent;
 import com.hazelcast.core.DistributedObjectListener;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.ITopic;
+import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import java.util.Collection;
+import java.util.concurrent.CountDownLatch;
 import org.junit.After;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.util.concurrent.CountDownLatch;
+import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
@@ -46,7 +50,7 @@ public class DistributedObjectListenerTest extends HazelcastTestSupport {
     @Test
     public void destroyedNotReceivedOnClient() throws Exception {
         HazelcastInstance instance = hazelcastFactory.newHazelcastInstance();
-        HazelcastInstance client = hazelcastFactory.newHazelcastClient();
+        final HazelcastInstance client = hazelcastFactory.newHazelcastClient();
         final CountDownLatch createdLatch = new CountDownLatch(1);
         final CountDownLatch destroyedLatch = new CountDownLatch(1);
         client.addDistributedObjectListener(new DistributedObjectListener() {
@@ -65,6 +69,13 @@ public class DistributedObjectListenerTest extends HazelcastTestSupport {
         assertOpenEventually(createdLatch, 10);
         topic.destroy();
         assertOpenEventually(destroyedLatch, 10);
+        assertTrueAllTheTime(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                Collection<DistributedObject> distributedObjects = client.getDistributedObjects();
+                assertTrue(distributedObjects.isEmpty());
+            }
+        }, 5);
     }
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/client/impl/client/AddDistributedObjectListenerRequest.java
+++ b/hazelcast/src/main/java/com/hazelcast/client/impl/client/AddDistributedObjectListenerRequest.java
@@ -76,7 +76,7 @@ public class AddDistributedObjectListenerRequest extends BaseClientAddListenerRe
         }
 
         PortableDistributedObjectEvent portableEvent = new PortableDistributedObjectEvent(
-                event.getEventType(), event.getDistributedObject().getName(), event.getServiceName());
+                event.getEventType(), (String) event.getObjectName(), event.getServiceName());
         endpoint.sendEvent(null, portableEvent, getCallId());
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AddDistributedObjectListenerMessageTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AddDistributedObjectListenerMessageTask.java
@@ -101,7 +101,7 @@ public class AddDistributedObjectListenerMessageTask
             return;
         }
 
-        String name = event.getDistributedObject().getName();
+        String name = (String) event.getObjectName();
         String serviceName = event.getServiceName();
         ClientMessage eventMessage =
                 ClientAddDistributedObjectListenerCodec.encodeDistributedObjectEvent(name,

--- a/hazelcast/src/main/java/com/hazelcast/core/DistributedObjectEvent.java
+++ b/hazelcast/src/main/java/com/hazelcast/core/DistributedObjectEvent.java
@@ -16,6 +16,8 @@
 
 package com.hazelcast.core;
 
+import com.hazelcast.spi.exception.DistributedObjectDestroyedException;
+
 /**
  * DistributedObjectEvent is fired when a {@link DistributedObject}
  * is created or destroyed cluster-wide.
@@ -25,22 +27,27 @@ package com.hazelcast.core;
  */
 public class DistributedObjectEvent {
 
+    protected DistributedObject distributedObject;
+
     private EventType eventType;
 
     private String serviceName;
 
-    private DistributedObject distributedObject;
+    private String objectName;
 
     /**
      * Constructs a DistributedObject Event.
      *
      * @param eventType         The event type as an enum {@link EventType} integer.
      * @param serviceName       The service name of the DistributedObject.
+     * @param objectName        The name of the DistributedObject.
      * @param distributedObject The DistributedObject for the event.
      */
-    public DistributedObjectEvent(EventType eventType, String serviceName, DistributedObject distributedObject) {
+    public DistributedObjectEvent(EventType eventType, String serviceName, String objectName,
+                                  DistributedObject distributedObject) {
         this.eventType = eventType;
         this.serviceName = serviceName;
+        this.objectName = objectName;
         this.distributedObject = distributedObject;
     }
 
@@ -79,15 +86,19 @@ public class DistributedObjectEvent {
      * @see DistributedObject#getName()
      */
     public Object getObjectName() {
-        return distributedObject.getName();
+        return objectName;
     }
 
     /**
      * Returns the DistributedObject instance.
      *
      * @return the DistributedObject instance
+     * @throws DistributedObjectDestroyedException if distributed object is destroyed.
      */
     public DistributedObject getDistributedObject() {
+        if (EventType.DESTROYED.equals(eventType)) {
+            throw new DistributedObjectDestroyedException(objectName + " destroyed!");
+        }
         return distributedObject;
     }
 
@@ -103,6 +114,7 @@ public class DistributedObjectEvent {
         return "DistributedObjectEvent{"
                 + "eventType=" + eventType
                 + ", serviceName='" + serviceName + '\''
+                + ", objectName='" + objectName + '\''
                 + ", distributedObject=" + distributedObject
                 + '}';
     }

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyEventProcessor.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyEventProcessor.java
@@ -28,19 +28,21 @@ final class ProxyEventProcessor implements StripedRunnable {
     private final Collection<DistributedObjectListener> listeners;
     private final DistributedObjectEvent.EventType type;
     private final String serviceName;
+    private final String objectName;
     private final DistributedObject object;
 
     ProxyEventProcessor(Collection<DistributedObjectListener> listeners, DistributedObjectEvent.EventType eventType,
-                        String serviceName, DistributedObject object) {
+                        String serviceName, String objectName, DistributedObject object) {
         this.listeners = listeners;
         this.type = eventType;
         this.serviceName = serviceName;
+        this.objectName = objectName;
         this.object = object;
     }
 
     @Override
     public void run() {
-        DistributedObjectEvent event = new DistributedObjectEvent(type, serviceName, object);
+        DistributedObjectEvent event = new DistributedObjectEvent(type, serviceName, objectName, object);
         for (DistributedObjectListener listener : listeners) {
             switch (type) {
                 case CREATED:

--- a/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java
+++ b/hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java
@@ -26,7 +26,6 @@ import com.hazelcast.spi.RemoteService;
 import com.hazelcast.spi.impl.eventservice.InternalEventService;
 import com.hazelcast.util.EmptyStatement;
 import com.hazelcast.util.ExceptionUtil;
-
 import java.util.Collection;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
@@ -205,7 +204,8 @@ public final class ProxyRegistry {
         }
 
         InternalEventService eventService = proxyService.nodeEngine.getEventService();
-        ProxyEventProcessor callback = new ProxyEventProcessor(proxyService.listeners.values(), CREATED, serviceName, proxy);
+        ProxyEventProcessor callback = new ProxyEventProcessor(proxyService.listeners.values(), CREATED, serviceName,
+                name, proxy);
         eventService.executeEventCallback(callback);
         if (publishEvent) {
             publish(new DistributedObjectEventPacket(CREATED, serviceName, name));
@@ -235,7 +235,8 @@ public final class ProxyRegistry {
             return;
         }
         InternalEventService eventService = proxyService.nodeEngine.getEventService();
-        ProxyEventProcessor callback = new ProxyEventProcessor(proxyService.listeners.values(), DESTROYED, serviceName, proxy);
+        ProxyEventProcessor callback = new ProxyEventProcessor(proxyService.listeners.values(), DESTROYED, serviceName,
+                name, proxy);
         eventService.executeEventCallback(callback);
         if (publishEvent) {
             publish(new DistributedObjectEventPacket(DESTROYED, serviceName, name));

--- a/hazelcast/src/test/java/com/hazelcast/core/DistributedObjectListenerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/core/DistributedObjectListenerTest.java
@@ -20,6 +20,7 @@ import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.QuickTest;
+import java.util.Collection;
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Test;
@@ -39,7 +40,7 @@ public class DistributedObjectListenerTest extends HazelcastTestSupport {
 
     @Test
     public void testDestroyJustAfterCreate() {
-        HazelcastInstance instance = Hazelcast.newHazelcastInstance();
+        final HazelcastInstance instance = Hazelcast.newHazelcastInstance();
         instance.addDistributedObjectListener(new EventCountListener());
         IMap<Object, Object> map = instance.getMap(randomString());
         map.destroy();
@@ -48,6 +49,8 @@ public class DistributedObjectListenerTest extends HazelcastTestSupport {
             public void run() throws Exception {
                 Assert.assertEquals(1, EventCountListener.createdCount.get());
                 Assert.assertEquals(1, EventCountListener.destroyedCount.get());
+                Collection<DistributedObject> distributedObjects = instance.getDistributedObjects();
+                Assert.assertTrue(distributedObjects.isEmpty());
             }
         };
         assertTrueEventually(task, 5);

